providers/stores/abstractFileStore.js

```diff
@@ -0,0 +1,126 @@
+const recursive = require('recursive-readdir')
+const { promisify } = require('util')
+const ResultCoordinates = require('../../lib/resultCoordinates')
+
+class AbstractFileStore {
+  constructor(options) {
+    this.options = options
+  }
+
+  async initialize() {}
+
+  /**
+   * Visit all of the files associated with the given coordinates.
+   *
+   * @param {EntityCoordinates} coordinates - Accepts partial coordinates.
+   * @returns The collection of results returned by the visitor
+   */
+  async list(coordinates, visitor) {
+    try {
+      const paths = await recursive(this._toStoragePathFromCoordinates(coordinates), ['.DS_Store'])
+      return await Promise.all(
+        paths.map(async path => {
+          if (!this._isValidPath(path)) return null
+          const data = await promisify(fs.readFile)(path)
+          return data ? visitor(JSON.parse(data)) : null
+        })
+      )
+    } catch (error) {
+      // If there is just no entry, that's fine, there is no content.
+      if (error.code === 'ENOENT') return []
+      throw error
+    }
+  }
+
+  /**
+   * Get and return the object at the given coordinates.
+   *
+   * @param {Coordinates} coordinates - The coordinates of the object to get
+   * @returns The loaded object
+   */
+  async get(coordinates) {
+    const filePath = await this._getFilePath(coordinates)
+    try {
+      const result = await promisify(fs.readFile)(filePath)
+      return JSON.parse(result)
+    } catch (error) {
+      if (error.code === 'ENOENT') return null
+      throw error
+    }
+  }
+
+  _isValidPath(entry) {
+    return AbstractFileStore.isInterestingCoordinates(this._toResultCoordinatesFromStoragePath(entry))
+  }
+
+  _toStoragePathFromCoordinates(coordinates) {
+    const result = AbstractFileStore.toStoragePathFromCoordinates(coordinates)
+    return path.join(this.options.location, result).replace(/\\/g, '/')
+  }
+
+  _toResultCoordinatesFromStoragePath(path) {
+    const trimmed = path.slice(this.options.location.length + 1)
+    return AbstractFileStore.toResultCoordinatesFromStoragePath(trimmed)
+  }
+
+  async _getFilePath(coordinates) {
+    const toolPath = this._toStoragePathFromCoordinates(coordinates)
+    if (coordinates.toolVersion) return toolPath + '.json'
+    const latest = await this._findLatest(toolPath)
+    if (!latest) return null
+    return path.join(toolPath, latest) + '.json'
+  }
+
+  _findLatest(filePath) {
+    return new Promise((resolve, reject) => {
+      fs.readdir(filePath, (error, list) => {
+        if (error) return reject(error)
+        const result = list.map(entry => (path.extname(entry) === '.json' ? path.basename(entry).slice(0, -5) : null))
+        resolve(utils.getLatestVersion(result.filter(e => e)))
+      })
+    })
+  }
+
+  // Static helper methods shared between path-based stores
+
+  static isInterestingCoordinates(coordinates) {
+    return ['gem', 'git', 'npm', 'maven', 'sourcearchive', 'nuget', 'pypi'].includes(coordinates.type)
```

Comment:
it would be handy if there was a way to read this list from https://github.com/clearlydefined/service/blob/master/schemas/definition.json#L20.

should we include `deadletter`?